Sources/PackageCollectionGenerator/PackageCollectionGenerate.swift

```diff
@@ -89,12 +89,37 @@ public struct PackageCollectionGenerate: ParsableCommand {
         let input = try jsonDecoder.decode(PackageCollectionGeneratorInput.self, from: Data(contentsOf: URL(fileURLWithPath: self.inputPath)))
         print("\(input)", verbose: self.verbose)
 
-        let githubPackageMetadataProvider = GitHubPackageMetadataProvider(authTokens: authTokens)
+        // Build the host-to-metadata-provider-mapping
+        var hostToProviderMapping: [String: PackageMetadataProvider] = [
+            "github.com": GitHubPackageMetadataProvider(authTokens: authTokens),
+            "gitlab.com": GitLabPackageMetadataProvider(authTokens: authTokens)
+        ]
```

Comment:
In my first try I implemented it this way, but afterwards I changed it to [String: String]. The main reason is the usability for the user. Currently, if the `input.json` contains a provider that is not implemented, the error-message will be `"Provider \(provider) for host \(host) is not implemented"`. If the provider is an `enum`, the decoding of the JSON will fail with a message like "Cannot initialize PackageMetadataProviderType from invalid String value bitbucket". This wouldn't guide the user what is really going wrong. Implementing the `init(from decoder: Decoder) throws`-method and throw a more descriptive error, like this
```
enum PackageMetadataProviderType: String {
	case github
	case gitlab
}

extension PackageMetadataProviderType: Codable {
	enum Errors: Error, Equatable, CustomDebugStringConvertible {
		case providerNotImplemented(String)
		
		var debugDescription: String {
			switch self {
				case let .providerNotImplemented(provider):
				return "Provider '\(provider)' is not implemented"
			}
		}
	}
	
	init(from decoder: Decoder) throws {
		let container = try decoder.singleValueContainer()
		let provider = try container.decode(String.self)
		switch provider {
			case "github":
				self = .github
			case "gitlab":
				self = .gitlab
			default:
				throw Errors.providerNotImplemented(provider)
		}
	}
}
```

improves the situation, as the user will at least get a message like "Provider 'bitbucket' is not implemented". This would be acceptable, but from my point of view it is not as good as the current message. And it doesn't simplify the implementation. Would you still prefer this implementation, or do you see another possibility that I didn't think of?

Another point where I wasn't sure about: In my current implementation, there is one instance of a `*MetaDataProvider`-class per Host, each one with it's own instance of `HTTPClient`. I'm not sure if this may cause any problems, if one instance would handle multiple hosts (I don't see any state in it on a first glance that could cause problems, but I haven't studied it enough to rule it out). Like, if there are three different hosts that make use of the gitlab-API - may it cause a problem, if all of them use the same `HTTPClient`-instance?

Regarding making `GitHubPackageMetadataProvider` as the default: I prefer to not make an internal assumption of a default-provider. It would lead to confusing and inconsistent errors, if it tries to use the GitHub-API for other providers. The current error-message `"Missing provider for host \(package.url.host ?? "invalid host"). Please specify it in the input-json in metadataProviders."` does not only give an error what is missing, but also offers a guidance where to add it. As the providers are already setup for `github.com`, it would not change the current behaviour.

Reply:
> In my first try I implemented it this way, but afterwards I changed it to [String: String]. The main reason is the usability for the user.

I see. What if we keep `metadataProviders` `[String: String]` in `PackageCollectionGeneratorInput` but introduce an intermediate, internal method that converts it to `[String: PackageMetadataProviderType]`? The method can throw error with user-friendly message when it comes across a provider that it doesn't recognize.

> Another point where I wasn't sure about: In my current implementation, there is one instance of a *MetaDataProvider-class per Host, each one with it's own instance of HTTPClient.

Yes, I don't think it's necessary to have one provider instance per host and that's why I suggest having `metadataProviderByType: [PackageMetadataProviderType: PackageMetadataProvider]` instead. I don't see a problem sharing the same provider instance across multiple hosts (`HTTPClient` is not tied to a host), but then I also don't think there would be so many different hosts that we would end up getting into trouble with one instance per host. 

I do prefer singleton by provider type though.

> I prefer to not make an internal assumption of a default-provider.

How about we add an option (e.g., `--default-metadata-provider`) that defaults to `nil` (none), but user can set it to one of `PackageMetadataProviderType` if they choose to?